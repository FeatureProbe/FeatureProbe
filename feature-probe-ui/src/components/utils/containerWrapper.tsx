import { I18NContainer } from 'hooks';

export default I18NContainer.Provider;
