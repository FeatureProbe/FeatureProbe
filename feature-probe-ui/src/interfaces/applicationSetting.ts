export interface IApplicationSetting {
  loginMode: string;
}
