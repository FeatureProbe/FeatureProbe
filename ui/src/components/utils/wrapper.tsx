import en_US from '../../locales/en-US.json';
import { IntlProvider } from 'react-intl';

export const IntlWrapper: React.FC = ({ children }) => {
  return (
    <IntlProvider locale="en" messages={en_US}>{children}</IntlProvider>
  );
};
