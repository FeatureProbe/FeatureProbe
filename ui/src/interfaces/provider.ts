/* eslint-disable @typescript-eslint/no-explicit-any */
import { ContainerProviderProps } from 'unstated-next';
import { ReactElement } from 'react';

interface IFn {
  (): any;
}
  
export interface IContainer {
  Provider: React.ComponentType<ContainerProviderProps<void>>;
  useContainer: IFn;
}

export interface IChildren {
  children: ReactElement
}