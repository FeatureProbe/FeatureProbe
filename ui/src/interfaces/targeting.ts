import { IPageable, ISort } from './toggle';

export interface IVariation {
  id?: string;
  name?: string;
  value?: string;
  description?: string;
  key?: string;
  text?: string;
  inputValue?: number;
  percentage?: string;
}

export interface IOption {
  key: string;
  text: string;
  value: string;
}

export interface ICondition {
  id?: string;
  type: string;
  subject?: string;
  predicate: string;
  objects?: string[];
  datetime?: string;
  timezone?: string;
}

export interface IServe {
  split?: number[];
  select?: number;
}

export interface IRule {
  id?: string;
  name: string;
  conditions: ICondition[];
  serve?: IServe;
  active?: boolean;
}

export interface IToggleInfo {
  name: string;
  key: string;
  tags: string[];
  desc: string;
  createdBy: string;
  createdTime: string;
  disabledServe: number;
  modifiedBy: string;
  modifiedTime: string;
  returnType: string;
  clientAvailability: boolean;
  variations: IVariation[];
  archived: boolean;
  permanent: boolean;
  useDays?: number;
}

export interface ITarget {
  rules: IRule[];
  variations: IVariation[];
  defaultServe: IServe;
  disabledServe: IServe;
}

export interface IContent {
  version: number;
  disabled: boolean;
  comment: string;
  content: ITarget;
  modifiedBy: string;
  modifiedTime: string;
  enableApproval: boolean;
  status: string;
  reviewers: string[];
  submitBy?: string;
  approvalBy?: string;
  approvalComment?: string;
  locked?: boolean;
  lockedTime?: string;
  publishTime?: string;
}

export interface IModifyInfo {
  modifiedBy?: string;
  modifiedTime?: string;
}

export interface IValues {
  count: number;
  value: string;
  deleted: boolean;
}

export interface IMetric {
  name: string;
  values: IValues[];
  lastChangeVersion?: number;
}

export interface IMetricContent {
  isAccess: boolean;
  metrics: IMetric[],
  summary: IValues[]
}

export interface IMetricParams {
  lastHours: string;
  metricType: string;
}

export interface IVersion {
  projectKey: string,
  comment: string,
  version: number,
  createdTime: string;
  createdBy: string;
  disabled: boolean;
  [key: string]: unknown; 
}

export interface ITargetingVersion extends IVersion {
  approvalBy: string;
  approvalStatus: string;
  approvalTime: string;
  approvalComment?: string;
  content: ITarget;
  environmentKey: string,
}

export interface ITargetingVersions {
  totalElements: number;
  totalPages: number;
  sort: ISort;
  first: boolean;
  last: boolean;
  numberOfElements: number;
  pageable: IPageable;
  size: number;
  content: ITargetingVersion[];
  number: number;
  empty: boolean;
}

export interface ITargetingVersionsByVersion {
  total: number;
  versions: ITargetingVersion[]
}

export interface IDictionary {
  key: string;
  createdTime: string;
  value: string;
  updatedTime: string;
}

export interface IDictionary {
  key: string;
  createdTime: string;
  value: string;
  updatedTime: string;
}

export interface IApprovalInfo {
  enableApproval: boolean;
  status: string;
  reviewers: string[];
  submitBy?: string;
  approvalBy?: string;
  approvalComment?: string;
  locked?: boolean;
  lockedTime?: string;
  publishTime?: string;
}

export interface ITargeting {
  disabled: boolean;
  content: ITarget;
}

export interface ITargetingParams {
  disabled: boolean;
  comment?: string;
  content: ITarget;
  reviewers?: string[];
}

export interface ITargetingDiff {
  currentDisabled: boolean;
  oldDisabled: boolean;
  currentContent: ITarget;
  oldContent: ITarget;
}