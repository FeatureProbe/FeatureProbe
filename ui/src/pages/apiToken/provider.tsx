import { ReactElement } from 'react';
import { createContainer } from 'unstated-next';
import { IContainer, IChildren } from 'interfaces/provider';
import { useTokenInfo, useReactHookForm } from './hooks';

export const tokenInfoContainer = createContainer(useTokenInfo);
export const hooksFormContainer = createContainer(useReactHookForm);

function compose(...containers: IContainer[]) {
  return function Component(props: IChildren) {
    return containers.reduceRight((children: ReactElement, Container: IContainer) => {
      return <Container.Provider>{children}</Container.Provider>;
    }, props.children);
  };
}
  
export const Provider = compose(tokenInfoContainer, hooksFormContainer);
  