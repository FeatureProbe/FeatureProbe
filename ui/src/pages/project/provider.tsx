import { ReactElement } from 'react';
import { createContainer } from 'unstated-next';
import { IContainer, IChildren } from 'interfaces/provider';
import { useProjectInfo, useEnvironmentInfo, useReactHookForm } from './hooks';

export const projectContainer: IContainer = createContainer(useProjectInfo);
export const environmentContainer: IContainer = createContainer(useEnvironmentInfo);
export const hooksFormContainer: IContainer = createContainer(useReactHookForm);

function compose(...containers: IContainer[]) {
  return function Component(props: IChildren) {
    return containers.reduceRight((children: ReactElement, Container: IContainer) => {
      return <Container.Provider>{children}</Container.Provider>;
    }, props.children);
  };
}
  
export const Provider = compose(projectContainer, environmentContainer, hooksFormContainer);
  