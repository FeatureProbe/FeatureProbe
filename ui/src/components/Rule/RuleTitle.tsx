import { SyntheticEvent, useCallback, useEffect, useState } from 'react';
import { Form, InputOnChangeData } from 'semantic-ui-react';
import { useIntl } from 'react-intl';
import PopConfirm from 'components/PopupConfirm'; 
import Icon from 'components/Icon';
import { IRule } from 'interfaces/targeting';
import { IContainer } from 'interfaces/provider';
import styles from './index.module.scss';

interface IProps {
  rule: IRule;
  index: number;
  disabled?: boolean;
  isHover: boolean;
  ruleContainer: IContainer;
  hooksFormContainer: IContainer;
}

const RuleTitle = (props: IProps) => {
  const { 
    rule, 
    index,
    disabled,
    isHover,
    ruleContainer,
    hooksFormContainer,
  } = props;

  const intl = useIntl();
  const [open, setOpen] = useState<boolean>(false);

  const {
    handleDeleteRule,
    handleInputRuleName,
  } = ruleContainer.useContainer();

  const {
    formState: { errors },
    unregister,
    clearErrors,
    getValues,
  } = hooksFormContainer.useContainer();

  useEffect(() => {
    if (!isHover) {
      setOpen(false);
    }
  }, [isHover]);
  
  const handleDelete = useCallback((e: SyntheticEvent, index: number, ruleId?: string) => {
    e.stopPropagation();

    for(const key in getValues()) {
      if (key.startsWith(`rule_${ruleId}_`)) {
        unregister(key);
        clearErrors(key);
      }
    }

    for(const key in errors) {
      if (key.startsWith(`rule_${ruleId}_`)) {
        clearErrors(key);
      }
    }
    handleDeleteRule(index);
  }, [unregister, clearErrors, getValues, handleDeleteRule, errors]);

  const handleInputClick = useCallback((e: SyntheticEvent) => {
    e.stopPropagation();
  }, []);

  const handleChange = useCallback((e: SyntheticEvent, data: InputOnChangeData) => {
    handleInputRuleName(index, data.value);
  }, [index, handleInputRuleName]);

	return (
    <div className={styles['rule-title']}>
      <Form.Group className={styles['rule-title-input']} widths='equal'>
        <Form.Field>
          <Form.Input
            className={styles['rule-input']}
            onClick={handleInputClick} 
            value={rule.name} 
            disabled={disabled}
            placeholder={`${intl.formatMessage({id: 'common.rule.text'})}${index + 1}`}
            onChange={async (e: SyntheticEvent, detail: InputOnChangeData) => {
              if (detail.value.length > 50 ) return;
              handleChange(e, detail);
            }}
          />
        </Form.Field>
      </Form.Group>
      {
        isHover && !disabled && (
          <span className={styles['rule-title-operation']}>
            <Icon customclass={styles['icon-drag']} type='drag' />
            <PopConfirm
              open={open}
              handleConfirm={(e: SyntheticEvent) => {
                e.stopPropagation();
                handleDelete(e, index, rule.id);
                setOpen(false);
              }}
              handleCancel={(e: SyntheticEvent) => {
                e.stopPropagation();
                setOpen(false);
              }}
              text={intl.formatMessage({id: 'targeting.rule.delete.text'})}
            >
              <Icon 
                type='archive' 
                customclass={styles['icon-archive']} 
                onClick={(e: SyntheticEvent) => {
                  setOpen(true);
                  e.stopPropagation();
                }}
              />
            </PopConfirm>
          </span>
        )
      }
    </div>
	);
};

export default RuleTitle;