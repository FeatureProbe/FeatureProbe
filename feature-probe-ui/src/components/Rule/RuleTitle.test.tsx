import { render, screen } from '@testing-library/react';
import '@testing-library/jest-dom';
import { DragDropContext, Droppable } from 'react-beautiful-dnd';
import userEvent from '@testing-library/user-event';
import { Form } from 'semantic-ui-react';
import React, { useEffect } from 'react';
import RuleTitle from '.';
import { IntlWrapper } from 'components/utils/wrapper';
import { hooksFormContainer, ruleContainer, variationContainer } from 'pages/targeting/provider';
import { IRule } from 'interfaces/targeting';

const RuleData = {
  id: 'test_id',
  name: '',
  serve: undefined,
  conditions: [],
  active: true,
};

const Wrapper: React.FC = ({ children }) => {
  return (
    <IntlWrapper>
      <hooksFormContainer.Provider>
        <ruleContainer.Provider>
          <variationContainer.Provider>
            <Form>
              <DragDropContext onDragEnd={jest.fn()}>
                <Droppable droppableId="droppable">
                  {(provided) => (
                    <div {...provided.droppableProps} ref={provided.innerRef}>
                      <div>
                        <UseHookWrapper>{children}</UseHookWrapper>
                      </div>
                      √ç{provided.placeholder}
                    </div>
                  )}
                </Droppable>
              </DragDropContext>
            </Form>
          </variationContainer.Provider>
        </ruleContainer.Provider>
      </hooksFormContainer.Provider>
    </IntlWrapper>
  );
};

const UseHookWrapper: React.FC = ({ children }) => {
  const { saveRules } = ruleContainer.useContainer();
  const { setValue } = hooksFormContainer.useContainer();

  useEffect(() => {
    const rule: IRule = {
      id: 'test_id',
      name: '',
      serve: undefined,
      conditions: [],
      active: true,
    };
    saveRules([rule, { ...rule, id: 'test_id2' }]);
    setValue('test_value_key', 'test_value');
  }, [saveRules, setValue]);

  return <>{children}</>;
};

it('RuleTitle snapshot', (done) => {
  const { asFragment, rerender } = render(
    <RuleTitle
      rule={RuleData}
      index={0}
      subjectOptions={[]}
      ruleContainer={ruleContainer}
      hooksFormContainer={hooksFormContainer}
    />,
    {
      wrapper: Wrapper,
    }
  );

  expect(asFragment()).toMatchSnapshot();

  rerender(
    <RuleTitle
      rule={RuleData}
      index={0}
      subjectOptions={[]}
      ruleContainer={ruleContainer}
      hooksFormContainer={hooksFormContainer}
      disabled={true}
    />
  );

  done();
});

test('RuleTitle opt', (done) => {
  (async () => {
    const { asFragment } = render(
      <RuleTitle
        rule={RuleData}
        index={0}
        subjectOptions={[]}
        ruleContainer={ruleContainer}
        hooksFormContainer={hooksFormContainer}
        variationContainer={variationContainer}
      />,
      {
        wrapper: Wrapper,
      }
    );

    const ruleTextbox = screen.getByPlaceholderText('Rule 1');
    await userEvent.hover(ruleTextbox);
    await userEvent.click(ruleTextbox);
    await userEvent.keyboard('test rule');
    await userEvent.type(ruleTextbox, 'test rule long long long long long long long long long long long long long' +
        ' long long long');

    const del = document.querySelector('span.icon-archive');
    del && (await userEvent.click(del));
    const cancel = screen.queryByText('Cancel');
    expect(cancel).not.toBeNull();

    cancel && (await userEvent.click(cancel));

    del && (await userEvent.click(del));
    const confirm = screen.queryByText('Confirm');
    expect(confirm).not.toBeNull();

    confirm && (await userEvent.click(confirm));

    expect(asFragment()).toMatchSnapshot();

    done();
  })();
});
