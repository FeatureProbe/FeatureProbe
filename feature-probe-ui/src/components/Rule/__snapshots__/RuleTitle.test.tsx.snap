// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RuleTitle opt 1`] = `
<DocumentFragment>
  <form
    class="ui form"
  >
    <div
      data-rbd-droppable-context-id="1"
      data-rbd-droppable-id="droppable"
    >
      <div>
        <div
          aria-describedby="rbd-hidden-text-1-hidden-text-3"
          data-rbd-drag-handle-context-id="1"
          data-rbd-drag-handle-draggable-id="rule_test_id"
          data-rbd-draggable-context-id="1"
          data-rbd-draggable-id="rule_test_id"
          draggable="false"
          role="button"
          tabindex="0"
        >
          <div
            class="rule"
          >
            <div
              class="accordion ui"
            >
              <div
                class="title rule-accordion-title"
              >
                <span
                  class="iconfont icon-angle-right icon-accordion"
                  customclass="icon-accordion"
                  type="angle-right"
                />
                <div
                  class="rule-title"
                >
                  <div
                    class="equal width fields rule-title-input"
                  >
                    <div
                      class="field"
                    >
                      <div
                        class="field rule-input"
                      >
                        <div
                          class="ui input"
                        >
                          <input
                            placeholder="Rule 1"
                            type="text"
                            value=""
                          />
                        </div>
                      </div>
                    </div>
                  </div>
                  <span
                    class="rule-title-operation"
                  >
                    <span
                      class="iconfont icon-drag icon-drag"
                      customclass="icon-drag"
                      type="drag"
                    />
                    <span
                      class="iconfont icon-archive icon-archive"
                      customclass="icon-archive"
                      type="archive"
                    />
                  </span>
                </div>
              </div>
              <div
                class="content"
              >
                <div
                  class="rule-content"
                >
                  <div
                    class="rule-add-one"
                    name="rule_test_id_add"
                  >
                    <div>
                      <button
                        class="ui secondary button rule-add-btn"
                        type="button"
                      >
                        <span
                          class="iconfont icon-add iconfont"
                          customclass="iconfont"
                          type="add"
                        />
                        Add
                      </button>
                    </div>
                  </div>
                  <div
                    class="serve"
                  >
                    <div
                      class="serve-select"
                    >
                      <span
                        class="serve-text"
                      >
                        <span
                          class="label-required"
                        >
                          *
                        </span>
                        Serve
                      </span>
                      <div
                        class="field serve-form-field"
                        style="width: 308px;"
                      >
                        <div
                          aria-expanded="false"
                          class="ui floating selection dropdown serve-dropdown"
                          name="rule_test_id_serve"
                          role="listbox"
                          style="width: 308px;"
                          tabindex="0"
                        >
                          <div
                            aria-atomic="true"
                            aria-live="polite"
                            class="divider text"
                            role="alert"
                          >
                            Please select serve
                          </div>
                          <span
                            class="iconfont icon-angle-down angle-down"
                            customclass="angle-down"
                            type="angle-down"
                          />
                          <div
                            class="menu transition"
                          >
                            <div
                              aria-checked="false"
                              aria-selected="true"
                              class="selected item"
                              role="option"
                              style="pointer-events: all;"
                            >
                              <span
                                class="text"
                              >
                                a percentage rollout
                              </span>
                            </div>
                          </div>
                        </div>
                      </div>
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
      Í
    </div>
  </form>
</DocumentFragment>
`;

exports[`RuleTitle snapshot 1`] = `
<DocumentFragment>
  <form
    class="ui form"
  >
    <div
      data-rbd-droppable-context-id="0"
      data-rbd-droppable-id="droppable"
    >
      <div>
        <div
          aria-describedby="rbd-hidden-text-0-hidden-text-0"
          data-rbd-drag-handle-context-id="0"
          data-rbd-drag-handle-draggable-id="rule_test_id"
          data-rbd-draggable-context-id="0"
          data-rbd-draggable-id="rule_test_id"
          draggable="false"
          role="button"
          tabindex="0"
        >
          <div
            class="rule"
          >
            <div
              class="accordion ui"
            >
              <div
                class="title rule-accordion-title"
              >
                <span
                  class="iconfont icon-angle-right icon-accordion"
                  customclass="icon-accordion"
                  type="angle-right"
                />
                <div
                  class="rule-title"
                >
                  <div
                    class="equal width fields rule-title-input"
                  >
                    <div
                      class="field"
                    >
                      <div
                        class="field rule-input"
                      >
                        <div
                          class="ui input"
                        >
                          <input
                            placeholder="Rule 1"
                            type="text"
                            value=""
                          />
                        </div>
                      </div>
                    </div>
                  </div>
                </div>
              </div>
              <div
                class="content"
              >
                <div
                  class="rule-content"
                >
                  <div
                    class="rule-add-one"
                    name="rule_test_id_add"
                  >
                    <div>
                      <button
                        class="ui secondary button rule-add-btn"
                        type="button"
                      >
                        <span
                          class="iconfont icon-add iconfont"
                          customclass="iconfont"
                          type="add"
                        />
                        Add
                      </button>
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
      Í
    </div>
  </form>
</DocumentFragment>
`;
